### ISSUE TITLE
Cross domain issues

### ISSUE THREAD

**Issue by limikael** [LearningLocker/learninglocker#447]
**Learning Locker version**
````
$ cat VERSION
1.1.0
````

**Steps to reproduce the problem**
Create any script that communicates with Learning Locker and run it in a web browser. Serve it from a different domain than where the Learning Locker is running. I used [TinCanJS](https://github.com/RusticiSoftware/TinCanJS) but I think any client library would do.

**Expected behaviour**
From what I have read xAPI should support cross domain requests, so I would expect Learning Locker to add the necessary headers to make this possible. For reference:

https://github.com/adlnet/xAPI-Spec/blob/master/xAPI.md#cors

My interpretation of that is that a xAPI server by default should do "as much as it can" to enable cross domain communication.

**Actual behaviour**
I'm getting the following error in Chrome:

XMLHttpRequest cannot load http://... Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials. 

And this in Firefox:

Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at [url]. This can be fixed by moving the resource to the same domain or enabling CORS. [url]

**Server information**
Standard installation with MongoDB.

**Additional information**
If you confirm it I can probably fix a pull request for it.

**Comment by ryansmith94** [comment]
Apologies for the delay. Still clearing some other tasks.

**Comment by ryansmith94** [comment]
Right finally getting to this. Really sorry that it's taken sooooo damn long!

**Comment by limikael** [comment]
Don't worry! You shouldn't really feel like you are held responsible, since we are not paying customers in that sense... :)

However, maybe an alternative approach would be to create a "hacking tutorial" as mentioned above, and maybe then I would be able to fix this myself...

It was weird when I tried to hack on it, it behaved as if the changes I made to any .php file just wouldn't make any difference whatsoever. As if I edited the wrong files or something, and I came to the conclusion that something was somehow cached on the server.

Would you know what could have caused this?

**Comment by ryansmith94** [comment]
I know you're not, I still don't like keeping you waiting though if it's something you really need. 

Yeah, I'm not too sure what that would look like, I'll maybe have a look at some other open source projects that do it. Do you have any examples? 

To be honest it may well have been the wrong file, but I couldn't say for sure.

**Comment by ee0pdt** [comment]
@limikael which php file were you editing?

**Comment by ryansmith94** [comment]
I can't seem to reproduce the issue, however I'm running one of the TinCan prototypes locally and I've setup my endpoint to be a local installation of LL via a virtual host. So I'm unsure if there is something dodgy there.

I can confirm that the `Access-Control-Allow-Credentials` header is not set to true in any of the responses' headers. However, `Access-Control-Allow-Headers` header is set to `X-EXPERIENCE-API-VERSION, AUTHORIZATION, CONTENT-TYPE` in all OPTIONS requests.

**Comment by ryansmith94** [comment]
Do you have something you can put up to reproduce the issue @limikael?

**Comment by limikael** [comment]
It is up now on:

http://limikael.altervista.org/swagmapdemo/apptest.html

It's using the LL instance at:

http://staging.tunapanda.org/learninglocker-1.2.1-beta/public/data/xAPI/

If you check the console there you see:

```
 loading from: http://limikael.altervista.org/swagmapdemo/xapiproxy
SwagMapController.js:24 loading state change, loading: true
SwagMapModel.js:149 loading swagmap: http://limikael.altervista.org/swagmapdemo/testmap.json
request.js:117 XHR finished loading: GET "http://limikael.altervista.org/swagmapdemo/testmap.json".
SwagMapModel.js:157 loaded... error: null
SwagMapController.js:24 loading state change, loading: false
SwagMapController.js:47 model loaded, length: 3
SwagMap.js:57 loading completion...
SwagMapController.js:24 loading state change, loading: true
apptest.html:1 XMLHttpRequest cannot load http://staging.tunapanda.org/learninglocker-1.2.1-beta/public/data/xAPI/sta…3Ali.mikael%40gmail.com%22%7D&activity=http%3A%2F%2Fsome%2Fxapi%2Factivity. Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials.
SwagItemModel.js:129 Uncaught TypeError: Cannot read property 'length' of undefined
apptest.html:1 XMLHttpRequest cannot load http://staging.tunapanda.org/learninglocker-1.2.1-beta/public/data/xAPI/sta…ael%40gmail.com%22%7D&activity=http%3A%2F%2Fsome%2Fother%2Fxapi%2Factivity. Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials.
SwagItemModel.js:129 Uncaught TypeError: Cannot read property 'length' of undefined
apptest.html:1 XMLHttpRequest cannot load http://staging.tunapanda.org/learninglocker-1.2.1-beta/public/data/xAPI/sta….mikael%40gmail.com%22%7D&activity=http%3A%2F%2Ftunapanda.org%2Fktouch%2F2. Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials.
SwagItemModel.js:129 Uncaught TypeError: Cannot read property 'length' of undefined
```

**Comment by limikael** [comment]
@ee0pdt: i don't remember which file it was... will see first if the above gives the guys any hints, if not i can try again to try and find which file it was...

**Comment by ryansmith94** [comment]
Hey @limikael, thanks for that. I think I've found the issue. I'll keep you posted.

**Comment by ryansmith94** [comment]
I might need you to pull a particular branch to test my fix if that's ok with you.

**Comment by limikael** [comment]
Cool! Tell me when and I'll pull!

**Comment by ryansmith94** [comment]
Will do, I'm just running the changes against the xAPI tests and LL tests.
Might take 10-20 minutes depending on issues.

**Comment by ryasmi** [comment]
Having some issues at the moment.

**Comment by ryansmith94** [comment]
Ok if you can pull the `issue/447` branch that would be great.

**Comment by limikael** [comment]
ok! trying that right away, will let you know...

**Comment by ryansmith94** [comment]
Brilliant thanks!

**Comment by ryansmith94** [comment]
Note that the issue/447 branch has came off 1.3.0 hence the upgrade requirements will apply. The requirements can be found on the [1.3.0 release](https://github.com/LearningLocker/learninglocker/releases/tag/v1.3.0).

**Comment by limikael** [comment]
Ok I will try this as soon as I get the chance... However it seems that I have misplaced my server keys, but I think I know where they are, so will be able to try it in a few hours...

**Comment by ryansmith94** [comment]
Ok no worries.

**Comment by limikael** [comment]
Damn I forgot to look for them on the other computer where I think I have them... I will test this tomorrow...

**Comment by ryansmith94** [comment]
No worries, thanks for letting me know :-)

**Comment by limikael** [comment]
They were not where I thought they were either... :( Ok working on getting hold of them, will test this as soon as I can!

**Comment by ryansmith94** [comment]
Alright thanks for keeping me posted :+1:

**Comment by limikael** [comment]
Hi,

Finally got a chance to test this, sorry for the delay...

Actually, the problem seems to persist... :(

You can test it here:

http://localhost/repo/swagmap/test/view/apptest.html

It uses the instance of LearningLocker that is here: http://staging.tunapanda.org/learninglocker.git/public/ which is using the issues/447 branch. If you check the console you will see the same error. In essence, this is the error message:

    Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials.

In my naive mind it would be as simple as adding that header to each response from the server? But maybe it's not that simple... Is this a header that needs to go in the request? And is the bug in that case in TinCanJS?

Anyway, checking that request in chrome shows these response headers:

```
Access-Control-Allow-Headers:X-EXPERIENCE-API-VERSION, AUTHORIZATION
Access-Control-Allow-Methods:POST, PUT, GET, DELETE, OPTIONS
Access-Control-Allow-Origin:http://limikael.altervista.org
Access-Control-Max-Age:3600
Cache-Control:no-cache
Connection:close
Content-Encoding:gzip
Content-Length:20
Content-Type:text/html
Date:Tue, 27 Jan 2015 14:08:50 GMT
Server:Apache/2.2.22 (Debian)
Set-Cookie:laravel_session=eyJpdiI6Imc0YWp3OWFJRndsV1ZRWGhtc0JCSUJFanJycUtMd2tsQjU4NHIzdlBQanM9IiwidmFsdWUiOiJUcVl5dkNYRnkzYW1yQytFV3dHZUo4QXFYZlhMVlwvSDM3RW5MdTluXC9SQlhMTWVwQzFVV09xcVFUU1VMWlU0Y0tDZGxxK1A1Q2VXdmtxZmxONTc5c3ZRPT0iLCJtYWMiOiI3ZTU0M2RmN2QxNTNiYzdmMGM3YzNjZGZmM2RmZjg3Mjg2ZjAwNjJhMDQ1YTRlMzFkMTYzMWI2MGU2NWJhNjA2In0%3D; expires=Tue, 27-Jan-2015 16:08:50 GMT; path=/; httponly
Vary:Accept-Encoding
X-Powered-By:PHP/5.4.4-14+deb7u14 
```

And these request headers:

```
Accept:*/*
Accept-Encoding:gzip, deflate, sdch
Accept-Language:en-US,en;q=0.8,nl;q=0.6,sv;q=0.4
Access-Control-Request-Headers:x-experience-api-version, authorization
Access-Control-Request-Method:GET
Cache-Control:no-cache
Connection:keep-alive
Host:staging.tunapanda.org
Origin:http://limikael.altervista.org
Pragma:no-cache
Referer:http://limikael.altervista.org/swagmapdemo/apptest.html
User-Agent:Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.91 Safari/537.36
```

Does that give any clue?

**Comment by ryansmith94** [comment]
That sucks. Giving up on that branch for now since I'm working on tidying up LL and I know it's going to be fixed in the `issue/tidyup` branch.

**Comment by limikael** [comment]
Ok! Thanks for the quick reply (as always :+1). Will be happy to test again!

**Comment by ryansmith94** [comment]
:smile: No worries :+1: Brilliant thank you!

**Comment by rklancer** [comment]
TL;DR: @limikael I think your original problem was due to using the npm-packaged "Node" version of TinCanJS instead of the Bower-packaged "browser" version. Probably you should use the bower version of TinCanJS (although it would be nice if TinCanJS offered a solution for Browserify users).

I can also confirm that I am able to send statements from a browser to a hosted Learning Locker instance (at lrs.learninglocker.net) without setting up any proxies or configuring the LL instance. However I'm using https://github.com/adlnet/xAPIWrapper in the client.

I investigated this a bit because at Concord Consortium we are considering using or extending LL for logging user events from in-browser educational simulation apps we create. At least for now we want to be able to ingest data regardless of what domain actually hosts the apps (eg if partners or other third parties host the apps). I encountered the same Chrome error as you when I tried using a custom (non-xAPI) log ingestor we host on Heroku; like LL the Heroku app sets the ACAO header to `*` to ingest data regardless of origin.

The solution in my case was to remove the `withCredentials` flag from the XHR object used in the client. (Though I had to peel the onion several layers back to find and fix the problem)

(Setting `<xhrObject>.withCredentials` to `true` allows the XHR request to present, and store, cookies from the request domain. This would be bad when ACAO is `*` because it would encourage the web service sending the ACAO: * header to think it had secured the session with a session cookie. However it would be trivial for any `evil.com`-hosted site to reuse that "secure" session)

It looks like @limikael's app is a Browserified app that uses the npm-packaged version of TinCanJS. However, that version seems to be specifically intended for use in Node; it uses the `xhr2` package (which simulates the XmlHttpRequest v.2 API for Node apps), which in turn delegates to the `request` package, which uses Node's `http`, which in the browser is emulated by `http-browserify` using the browser's original xhr!

The problem is at the bottom of that stack, where `http-browserify` sets `withCredentials` to `true` if it is undefined: https://github.com/substack/http-browserify/blob/17b2990010ebd39461d1117c1e2c50c25eab869f/lib/request.js#L18-L20

See https://github.com/request/request/issues/986 where the entirely reasonable conclusion is that `request` is for Node and doesn't have to/can't support CORS (though someone follows up by mentioning Browserify)

As for TinCanJS, see the README about Node.js vs. browser usage, and https://github.com/RusticiSoftware/TinCanJS/tree/master/src/Environment -- note that the browser.js doesn't seem to touch `withCredentials` or use the `xhr2` package (leading me to conclude that the bower/browser version would work just fine with the wildcard ACAO header)

I do think it's unfortunate that you're being forced to switch to Bower or to concat a file instead of using npm+Browserify for your use case, though.

**Comment by ryansmith94** [comment]
Thanks for the input @rklancer. @limikael please read @rklancer's comment above and provide feedback regarding this.

**Comment by ryansmith94** [comment]
Closing this issue until told that @rklancer's comment does not resolve this.

**Comment by limikael** [comment]
Ok finally I had a chance to test this...

It seems to work fine when using the web adapted version rather than the node.js adapted one.

Thanks for testing!

**Comment by ryansmith94** [comment]
Nice one! That's great news! Thanks for letting us know :smile:

**Comment by brianjmiller** [comment]
FWIW, TinCanJS ships both the Node.js built file and the browser built file via `npm`, so it is a matter of pointing at the proper file for the environment where you intend to use the client. For those using TinCanJS from Node.js you should be loading `build/tincan-node.js` which you can do just by using `require("tincanjs");`. For those using TinCanJS from a browser you want to load either `build/tincan.js` (if you plan to minify yourself, or `build/tincan-min.js` if you want to use the already minified version. I'm not super familiar with Browserify but it should be possible to load the browser based file using a shim or some sort of configuration to point to the correct file.

**Comment by FlorianTolk** [comment]
I am using [https://github.com/adlnet/xAPIWrapper](https://github.com/adlnet/xAPIWrapper) but am getting the same error claiming the Cross Origin Policy won't let me access it.
